lucet-runtime/lucet-runtime-internals/src/alloc/mod.rs

```diff
@@ -154,25 +154,27 @@ impl Alloc {
         // the above makes sure this expression does not underflow
         let guard_remaining = self.heap_inaccessible_size - expand_pagealigned as usize;
 
-        let heap_spec = module.heap_spec();
-        // The compiler specifies how much guard (memory which traps on access) must be beyond the
-        // end of the accessible memory. We cannot perform an expansion that would make this region
-        // smaller than the compiler expected it to be.
-        if guard_remaining < heap_spec.guard_size as usize {
-            bail_limits_exceeded!("expansion would leave guard memory too small");
-        }
+        if let Some(heap_spec) = module.heap_spec() {
+            // The compiler specifies how much guard (memory which traps on access) must be beyond the
+            // end of the accessible memory. We cannot perform an expansion that would make this region
+            // smaller than the compiler expected it to be.
+            if guard_remaining < heap_spec.guard_size as usize {
+                bail_limits_exceeded!("expansion would leave guard memory too small");
+            }
 
-        // The compiler indicates that the module has specified a maximum memory size. Don't let
-        // the heap expand beyond that:
-        if let Some(max_size) = heap_spec.max_size {
-            if self.heap_accessible_size + expand_pagealigned as usize > max_size as usize {
-                bail_limits_exceeded!(
-                    "expansion would exceed module-specified heap limit: {:?}",
-                    max_size
-                );
+            // The compiler indicates that the module has specified a maximum memory size. Don't let
+            // the heap expand beyond that:
+            if let Some(max_size) = heap_spec.max_size {
+                if self.heap_accessible_size + expand_pagealigned as usize > max_size as usize {
+                    bail_limits_exceeded!(
+                        "expansion would exceed module-specified heap limit: {:?}",
+                        max_size
+                    );
+                }
             }
+        } else {
+            bail_limits_exceeded!("no heap to expand");
```

Comment:
Agreed, I made it a distinct error type. However, since we can't distinguish between a call to `lucet_vmctx_grow_memory` that is directly from guest code, or from a host call using the C API, I'm not going to make it die on that yet.